### PR TITLE
build: update go version to 1.21 from 1.20

### PR DESCRIPTION
**Pull request by subhamkrai**
**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -27,7 +27,7 @@ jobs:
 
       - uses: actions/setup-go@v4
         with:
-          go-version: "1.20"
+          go-version: "1.21"
```
Comment (parth-gr) [inline .github/workflows/build.yml:30]: linux build is failing

Comment (subhamkrai) [reply .github/workflows/build.yml:30]: fixed, in makeFile

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/rook/rook
 
-go 1.20
+go 1.21
```
Comment (travisn) [inline go.mod:3]: I don't think we can update this until we drop 1.20 support.

Comment (subhamkrai) [reply go.mod:3]: I think an earlier upgrade is better for us, or sometime soon some cve will be open to fix the golang version, do we need to wait for something before upgrading?

**Review by travisn — APPROVED** [review]

**Pull request merged**